Add conversion doc (#55)
openvinotoolkit/mediapipe: mediapipe/calculators/ovms/calculators.md

|:---|:---|:---|:---|:---|
|input_stream: "a"|input|none|ov::Tensor|a|
|output_stream: "OVTENSOR:b"|output|OVTENSOR|ov::Tensor|b|
|output_stream: "OVTENSORS:b"|output|OVTENSORS|std::vector\<ov::Tensor\>|b|
|output_stream: "TENSOR:b"|output|TENSOR|mediapipe::Tensor|b|
|input_stream: "TENSORS:b"|input|TENSORS|std::vector\<mediapipe::Tensor\>|b|

In case of missing tag calculator assumes that the packet type is `ov::Tensor'.

> # How to reuse existing graph from MediaPipe in OpenVINO Model Server

Reviewer: ```suggestion
# How to deploy existing graphs from MediaPipe framework with OpenVINO inference calculator
```

## How to get models used in MediaPipe demos

> To get the model used in MediaPipe demo you can either trigger build target that depends upon that model and then search in bazel cache or download directly from locations below.

Reviewer: ```suggestion
When you build mediapipe applications or solutions, typically the bazel configuration would download the needed models as a dependency. When the graph is to be deployed via the OpenVINO Model Server or when the mediapipe application would use OpenVINO Model Server as the inference executor, the models needs to be stored in the [models repository](https://docs.openvino.ai/2023.2/ovms_docs_models_repository.html).
That way you can take advantage of the model versioning feature and store the models on the local or the cloud storage. The OpenVINO calculator is using as a parameter the path to the [config.json](https://docs.openvino.ai/2023.2/ovms_docs_serving_model.html#serving-multiple-models) file with models configuration with the specific model name.

To get the model used in MediaPipe demo you can either trigger build target that depends upon that model and then search in bazel cache or download directly from locations below.
```

* https://storage.googleapis.com/mediapipe-models/
* https://storage.googleapis.com/mediapipe-assets/

> ## How to prepare OpenVINO Model Server deployment with Mediapipe

Reviewer: I think this section might be more suitable for model_server repo

```
servables/
├── config.json
└── dummyAdd
    ├── add_two_inputs_model
    │   └── 1
    │       ├── add.bin
    │       └── add.xml
    ├── dummy
    │   └── 1
    │       ├── dummy.bin
    │       └── dummy.xml
    ├── graph.pbtxt
    └── subconfig.json
```
and config.json:
```
{
  "model_config_list": [],
  "mediapipe_config_list": [
    {
      "name":"dummyAdd"
    }
  ]
}
```
and the subconfig.json:
```
{
  "model_config_list": [
    {
      "config": {
        "name": "dummy",
        "base_path": "dummy"
      }
    },
    {
      "config": {
        "name": "add",
        "base_path": "add_two_inputs_model"
      }
    }
  ]
}
```
In both cases `servables` directory will be mounted to OVMS container. You can find more details about OVMS configuration in [documentation](https://docs.openvino.ai/2023.1/ovms_docs_serving_model.html#serving-multiple-models).

*Note*: base paths in config.json are relative to the file path of config.json.
## How to adjust existing graphs to perform inference with OpenVINO Model Server

> Now we will show steps that to convert existing graph to use OpenVINO for inference.

Reviewer: ```suggestion
Below are presented steps to adjust existing graph using TensorFlow or TensorFlowLite calculators to use OpenVINO as the inference engine. That way the inference execution can be optimized with preserving the overall results graph structure.
```

> ### 1. First step is *optional*.

Reviewer: ```suggestion
### 1. Identify all used subgraphs with inference calculators. This steps is not needed if there are no subgraphs.
```

```
grep -R -n "register_as = \"HolisticLandmarkCpu"
```
We will find that in using bazel `mediapipe_simple_subgraph` function another `pbtxt` file was registered as a graph. Since in that file there is no inference calculator we need to repeat the procedure until we find all inference calculators used directly or indirectly using subgraphs.
### 2. We need to start with basic replacement of inference calculator.
Existing configuration could look like:
```
node {
  calculator: "HandLandmarkModelLoader"
  input_side_packet: "MODEL_COMPLEXITY:model_complexity"
  output_side_packet: "MODEL:model"
}
node {
  calculator: "InferenceCalculator"
  input_side_packet: "MODEL:model"
  input_stream: "TENSORS:input_tensor"
  output_stream: "TENSORS:output_tensors"
  options: {
    [mediapipe.InferenceCalculatorOptions.ext] {
      model_path: "mediapipe/modules/holistic_landmark/hand_recrop.tflite"
      delegate {
        xnnpack {}
      }
    }
  }
}
```

> This tells us which model is used (hand_recrop) and what type of packets are send to inference calculator (vector\<mediapipe::Tensor\>). We also need information what are model names inputs. This could be checked f.e. using OVMS logs or metadata request calls. With that information we would replace that part of a graph with:

Reviewer: This is not clear how the ovms logs are related to the model metadata.

```
node {
  calculator: "OpenVINOModelServerSessionCalculator"
  output_side_packet: "SESSION:session"
  node_options: {
    [type.googleapis.com / mediapipe.OpenVINOModelServerSessionCalculatorOptions]: {
      servable_name: "hand_recrop"
      servable_version: "1"
    }
  }
}
node {
  calculator: "OpenVINOInferenceCalculator"
  input_side_packet: "SESSION:session"
  input_stream: "TENSORS:initial_crop_tensor"
  output_stream: "TENSORS:landmark_tensors"
  node_options: {
    [type.googleapis.com / mediapipe.OpenVINOInferenceCalculatorOptions]: {
          tag_to_input_tensor_names {
            key: "TENSORS"
            value: "input_1"
          }
          tag_to_output_tensor_names {
            key: "TENSORS"
            value: "output_crop"
          }
        }
  }
}
```
In `OpenVINOModelServerSessionCalculator` we set `servable_name` with the model's name we found earlier. In `OpenVINOInferenceCalculator` we set input & output tags names to start with `TENSORS`. We then need to map out those tags to actual model names in `mediapipe.OpenVINOInferenceCalculatorOptions` `tag_to_input_tensor_names` and `tag_to_output_tensor_names` fields.

> ### 3. Third step is *optional* but may be required if model has multiple inputs/outputs.

Reviewer: ```suggestion
### 3. Third step is required if model has multiple inputs or outputs.
```

In case of missing tag calculator assumes that the packet type is `ov::Tensor'.

## How to adjust existing graphs to perform inference with OpenVINO Model Server
To make already prepared graphs use OpenVINO Model Server for inferences there are following steps involved:
1. Prepare OVMS servables repository with [all required models for graph](https://docs.openvino.ai/2023.1/ovms_docs_serving_model.html#serving-multiple-models). Unless you plan to reuse models in several graphs, it is recommended to use following structure:

> ```

Reviewer: it is not clear what kind of model reuse you talk about. Maybe for clarity it might be better just to describe one example as recommendation?

servables/
├── config.json

> └── dummyAdd

Reviewer: maybe some more meaningful names would be easier to follow...

    ├── add_two_inputs_model
    │   └── 1
    │       ├── add.bin
    │       └── add.xml
    ├── dummy
    │   └── 1
    │       ├── dummy.bin
    │       └── dummy.xml
    ├── graph.pbtxt
    └── subconfig.json
```

> Where `servables` directory will be mounted to OVMS container. If you plan to reuse models with the same configuration it is better to keep all models configuration in main config.json file.

Reviewer: This contradicts with the recommendation just mentioned

2. Prepare OVMS configuration files. In main config file setup MediaPipe graph:

> ![MainConfig](MainConfig.png)

Reviewer: text in images form is very inconvenient to use...
